packages/connectors-lib/src/distributed-lock.js

```diff
@@ -0,0 +1,103 @@
+import Redlock from 'redlock'
+import Redis from 'ioredis'
+import db from 'debug'
+const debug = db('connectors:distributed-lock')
+
+/**
+ * Allows for the creation of a distributed lock using redis/redlock
+ */
+export class DistributedLock {
+  constructor (lockName, ttlMs) {
+    this._lockName = lockName
+    this._ttl = ttlMs
+    this._lock = null
+    this._extendTimer = null
+    this._redlock = null
+  }
+
+  /**
+   * Initialise redlock/redis
+   *
+   * @returns {Promise<void>}
+   * @private
+   */
+  async _initialiseRedlock () {
+    this._redlock = new Redlock([
+      new Redis({
+        host: process.env.REDIS_HOST || 'localhost',
+        port: process.env.REDIS_PORT || 6379,
+        ...(process.env.REDIS_PASSWORD && {
+          password: process.env.REDIS_PASSWORD,
+          tls: {}
+        })
+      })
+    ])
+    this._redlock.on('clientError', console.error)
+  }
+
+  /**
+   * Attempt to obtain a lock.  This method will wait up until maxWaitSeconds to obtain a lock.  If a lock cannot be obtained within this time
+   * then an exception will be thrown.  Set maxWaitSeconds to -1 to wait forever, set maxWaitSeconds to 0 to throw immediately.  A positive value
+   * defines the number of seconds to wait before throwing
+   *
+   * @param {number} maxWaitSeconds the time to wait to obtain a lock
+   * @returns {Promise<void>}
+   */
+  async obtain ({ maxWaitSeconds = -1 } = {}) {
+    await this._initialiseRedlock()
+    const invocationTime = new Date()
+    let elapsed = 0
+    do {
+      debug('Attempting to obtain lock for %s', this._lockName)
+      try {
+        this._lock = await this._redlock.lock(this._lockName, this._ttl)
+        elapsed = (new Date() - invocationTime) / 1000
+      } catch (e) {
+        // Ignored
+      }
+      // eslint-disable-next-line no-unmodified-loop-condition
+    } while (!this._lock && (maxWaitSeconds < 0 || elapsed < maxWaitSeconds))
```

Comment:
Are you hammering on the lock here? Should this have a wait in it, or is it built into the lock?

Reply:
Yes, redlock has an internal wait/attempt mechanism each time you attempt to acquire the lock.